File: Servisno-orijentisane Arhitekture/Domaci/03. gRPC/Services/EmailService.cs
using Grpc.Core;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace _03._gRPC
{
    public class EmailService : Email.EmailBase
    {
        private readonly ILogger<EmailService> _logger;
        private static Dictionary<String, User> _users = null;
        private Dictionary<String, User> Users
        {
            get
            {
                if (_users == null)
                    _users = new Dictionary<string, User>();
                return _users;
            }
        }
        public EmailService(ILogger<EmailService> logger)
        {
            _logger = logger;
        }

        public override Task<Response> Register(Credentials credentials, ServerCallContext context)
        {
            Response ret = new Response();
            if (credentials == null || String.IsNullOrEmpty(credentials.Email) || String.IsNullOrEmpty(credentials.Password))
            {
                ret.Code = 401;
                ret.Message = "Incorrect email or password!";
            }
            else if (Users.ContainsKey(credentials.Email))
            {
                ret.Code = 401;
                ret.Message = "Email already registered!";
            }
            else {
                Users.Add(credentials.Email, new User()
                {
                    Credentials = credentials,
                    Unread = new List<Content>()
                });
                ret.Code = 200;
                ret.Message = "User successfully registered!";
            }

            return Task.FromResult(ret);
        }

        public override Task<Response> SendEmail(Content email, ServerCallContext context)
        {
            Response ret = new Response();

            if (!Users.ContainsKey(email.FromUser))
            {
                ret.Code = 401;
                ret.Message = "Sender not registered!";
            }
            else if (!Users.ContainsKey(email.ToUser))
            {
                ret.Code = 401;
                ret.Message = "Receiver not registered!";
            }
            else
            {
                Users[email.ToUser].Unread.Add(email);
                ret.Code = 200;
                ret.Message = "Email sent successfully!";
            }

            return Task.FromResult(ret);
        }

        public override async Task CheckMailbox(Credentials credentials, IServerStreamWriter<Content> responseStream, ServerCallContext context)
        {
            if (credentials != null && !String.IsNullOrEmpty(credentials.Email) && !String.IsNullOrEmpty(credentials.Password))
            {
                User u = null;
                if (Users.TryGetValue(credentials.Email, out u) && u.Credentials.Password == credentials.Password)
                {
                    foreach (Content unread in u.Unread)
                        await responseStream.WriteAsync(unread);
                    u.Unread.Clear();
                }
            }
        }
    }
}

File: Servisno-orijentisane Arhitekture/Domaci/03. gRPC/Node/services/email-service.js
const PROTO_PATH = __dirname + '/../../Protos/email.proto';

const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');

const packageDefinition = protoLoader.loadSync(
    PROTO_PATH, {
        keepCase: true,
        longs: String,
        enums: String,
        defaults: true,
        oneofs: true
    }
);

const protoDescriptor = grpc.loadPackageDefinition(packageDefinition);

const addEmailService = (server) => {
    server.addService(protoDescriptor.email.Email.service, {
        register: register,
        sendEmail: sendEmail,
        checkMailbox: checkMailbox
    });
};

module.exports = {
    addEmailService
};


// Implementation
const users = {};

function register(call, callback) {
    if (!call.request.email || !call.request.password) {
        callback(null, {
            code: 401,
            message: "Wrong credentials!"
        });
    } else if (users[call.request.email] != null) {
        callback(null, {
            code: 401,
            message: "Email already registered!"
        });
    } else {
        users[call.request.email] = {
            credentials: call.request,
            unread: []
        };

        callback(null, {
            code: 200,
            message: "User successfully registered!"
        });
    }
};

function sendEmail(call, callback) {
    if (!call.request.to_user || !call.request.from_user || !call.request.message) {
        callback(null, {
            code: 401,
            message: "Wrong email format!"
        });
    } else if (!users[call.request.to_user]) {
        callback(null, {
            code: 401,
            message: "Receiver not registered!"
        });
    } else if (!users[call.request.from_user]) {
        callback(null, {
            code: 401,
            message: "Sender not registered!"
        });
    } else {
        users[call.request.to_user].unread.push(call.request);

        callback(null, {
            code: 200,
            message: "Email sent successfully!"
        });
    }
};

function checkMailbox(call) {
    if (!call.request.email || !call.request.password)
        return call.end();

    const u = users[call.request.email];
    if (!u || u.credentials.password != call.request.password)
        return call.end();

    u.unread.forEach(email => call.write(email));
    return call.end();
};